### ISSUE TITLE
Confusing 404s when combining multiple rest parameters, slugs and param matchers 

### ISSUE THREAD

**Issue by arnolicious** [sveltejs/kit#13106]
### Describe the bug

Hi everyone,
Honestly I'm not quite sure where to start, I've hit a use case where the routing isn't quite working how I think it should be working, so I would very tentatively call it a bug.

## Scenario
The page I'm concerned about is at the path `event/:id` and below.
1. There can be any number of previous path segments before hand, thus I added a `[...path]` before.
2. The "event" string could actually be multiple different strings, `event | project | job`, so i have a param matcher for that
3. The `id` being a slug is quite self explanatory
4. Then the main page of the `event` is actually under `event/:id/overview`, so I've got that redirect in `event/[id]/+page.ts`
5. Now if a path like `event/:id/invalid` gets hit, I want to display an error *inside* of the event page, so `event/[id]/+error.svelte`, thus the `event/:id/[...catchall]/+page.ts`.

Entire Structure:
![image](https://github.com/user-attachments/assets/27397431-c62b-4377-be44-85840538e146)

## Expected Behaviour
Sveltekit hits the `event/:id/[...catchall]/+page.ts`, throws the 404 that gets caught in the  `event/[id]/+error.svelte`.

## Actual Behaviour
The entire route gets 404-ed and the generic 404 message gets displayed.

## Now it does actually work as expected, if either one of these modifications is made:
- `/[id]/ -> /1/` : If the `id` is not a slug, but a literal `1` (for example to match the hrefs), it does work correctly (the red outline around the 404 message shows that it's the correct `+error.svelte`
- `/[event=eventSlug]/ -> /event/` : If the `event` segment does not have a param matcher, it also works correctly

### Addendum
I also think I experienced more weirdness, where the routing wouldn't behave the same when built on windows compare to linux, where on my local windows machine it would work as expected and on linux rest parameters would get hit instead of other explicit paths on the same level, but I haven't really been able to reproduce that so far.

### Reproduction

https://stackblitz.com/edit/sveltejs-kit-template-default-win19u?file=src%2Fparams%2FidParam.ts

### System Info

```Shell
Stackblitz:
  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @sveltejs/adapter-auto: ^3.0.0 => 3.3.1 
    @sveltejs/kit: ^2.9.0 => 2.9.0 
    @sveltejs/vite-plugin-svelte: ^5.0.0 => 5.0.1 
    svelte: ^5.0.0 => 5.5.2 
    vite: ^6.0.0 => 6.0.2

Local Machine:
  System:
    OS: Windows 11 10.0.26120
    CPU: (16) x64 AMD Ryzen 7 5800X3D 8-Core Processor
    Memory: 11.15 GB / 31.91 GB
  Binaries:
    Node: 20.9.0 - C:\Program Files\nodejs\node.EXE
    npm: 10.2.5 - C:\Program Files\nodejs\npm.CMD
    pnpm: 8.10.2 - ~\AppData\Local\pnpm\pnpm.EXE
    bun: 1.1.38 - ~\.bun\bin\bun.EXE
  Browsers:
    Edge: Chromium (131.0.2903.36)
    Internet Explorer: 11.0.26100.1
  npmPackages:
    @sveltejs/adapter-auto: ^3.0.0 => 3.3.1
    @sveltejs/kit: ^2.9.0 => 2.9.0
    @sveltejs/vite-plugin-svelte: ^5.0.0 => 5.0.1
    svelte: ^5.0.0 => 5.5.2
    vite: ^6.0.0 => 6.0.2
```

### Severity

blocking all usage of SvelteKit

**Comment by arnolicious** [comment]
So I've tried to look into the kit codebase a little bit, and I've noticed a suspicious difference in the sorting of the generated routes.

I logged out `sort_routes(routes)` in this file: https://github.com/sveltejs/kit/blob/main/packages/kit/src/core/sync/create_manifest_data/index.js#L464

Using the Reproduction Repo from above, these were the outputs:

## Working as expected:
Route setup:
```
[...path]/event/[id]/...
```  
[correct-routes.txt](https://github.com/user-attachments/files/18497098/correct-routes.txt)

## Bug occurs:
Route setup:
```
[...path]/[event=eventSlug]/[id]/...
```  
[buggy-routes.txt](https://github.com/user-attachments/files/18497097/buggy-routes.txt)


## Suspicion
Besides the changes that are due to the addition of the matcher, everything seems to be the same **except** where the `id: '/[...path]'` route gets sorted.

In the buggy scenario it gets sorted second, right after the root route, in the working scenario it gets sorted last.

Could this be the cause of this?

**Comment by arnolicious** [comment]
Ok, nevermind my last comment, that didn't seem to be it.

However, I think this is more like it:

The failed navigation in the buggy example occurs because the `exec` function here fails:
https://github.com/sveltejs/kit/blob/main/packages/kit/src/utils/routing.js#L139

While the example without the param matcher does not fail.

That difference comes from the MatchArray (`match`) parameter that the function receives:
```
Correct Scenario:
['/calendar/20204/test/1/invalid', 'calendar/20204', '1', 'invalid', index: 0, input: '/calendar/20204/test/1/invalid', groups: undefined]

Buggy Scenario:
['/calendar/20204/event/1/invalid', 'calendar/20204/event', '1', 'invalid', index: 0, input: '/calendar/20204/event/1/invalid', groups: undefined]
``` 
Specifically the problem is that the `event` part is not a separate item in that array, which causes the param matcher to fail, since it tries to validate the `"1"` instead of `"event"`

I think the array should look like this:
```
['/calendar/20204/event/1/invalid', 'calendar/20204', 'event', '1', 'invalid', index: 0, input: '/calendar/20204/event/1/invalid', groups: undefined]
```

## Conclusion
The problems stems from the regex pattern generated in https://github.com/sveltejs/kit/blob/main/packages/kit/src/utils/routing.js#L9
which results in different match patterns.

Basically, the `[event=eventSlug]` part of the path gets "eaten" by the `[...path]` part, and then it's missing in the match array.

Unfortunately i'm not smart enough to solve this regex problem, or rather I'm not sure if this is solvable as is?
Maybe you'd need to "inject" the actual param matching functions in here somehow?

Anyways, I'd be really thankful if someone had a fix or a suggestion on how to make this scenario work a bit better for me :)
